actually fixes blob spore pathfinding

MNarath1 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
In https://github.com/BeeStation/BeeStation-Hornet/pull/5967 i tried to fix blob pathfinding by simply replacing the normal pathfinding with jps pathfinding turned out trough there was more to  this than just that and some test cases that i did not think of at that time revealed problems that i am gonna adress with this PR.
closes https://github.com/BeeStation/BeeStation-Hornet/issues/6098
Edit:
The fallback for pathfinding works by making the assumtion that we have something blocking our way so we step from the target towards the origin until we find a turf that either is a ~~turf/closed~~ a turf with density true and no matching passflag or has an object on it that qualfies as a blocker when we find it we step 1 further and if that one is free so no blockers we try to pathfind towards that if we still fail the same procedure goes again until we stepped completly back towards the origin if we get there then pathfinding truely has failed.
~~Important Note trough not all possible blockers are adressed so there still might be scenarios that would have it fail (cause i can't just assume all machines as blockers cause pipes are machines too etc.)~~
This is a non issue after going away from checking their typpath and instead check if they are dense and if the passflags don't match that way we can adress any possible blocker.

## Why It's Good For The Game
FIX

## Changelog
:cl:
fix: fixed blob spore pathing
tweak: blob spores will keep hunting target mobs until those are out of range
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

MNarath1 committed:
just some stuff

MNarath1 committed:
almost done just a bit more

MNarath1 committed:
additional changes

MNarath1 committed:
how tf did i do that

MNarath1 committed:
consistancy

MNarath1 committed:
comments

MNarath1 commented:
also yes the fallback is not perfect but i feel looking for the blocker is a good enough compromise to find out if we can try to pathfind again and should reduce the actually needed calls for get_path_to in this proc a LOT and thats good considering its the most expensive proc in here.

MNarath1 committed:
forgot comment

Penwin0 commented:
Hero

MNarath1 committed:
comment change

MNarath1 committed:
small change

MNarath1 committed:
forgot about tables whops

MNarath1 commented on code/modules/antagonists/blob/blob_mobs.dm:
```diff
@@ -208,16 +209,45 @@
+			for(var/i in 1 to ln) //calling get_path_to every time is quite taxing lets see if we can find whatever blocks us
+				target_new = get_step(target_new,  get_dir(target_new, src)) //step towards the origin until we find the blocker then 1 further
+				ln--
+				if(istype(target_new,/turf/closed) || (locate(/obj/structure/window) in target_new) || (locate(/obj/machinery/door) in target_new) || (locate(/obj/structure/table) in target_new) || (locate(/obj/mecha) in target_new)) //we check for possible things that could block us
```
If anyone has a better idea how to do this part i am very much open to suggestions

MNarath1 replied:
i mean line 237

Archanial replied:
for var/stuff in turf.contents if istype in list

MNarath1 replied:
> for var/stuff in turf.contents if istype in list

Already tried that by benchmarking it for some reason it seems to be faster to do this with locate but that makes no sense at all honestly

MNarath1 committed:
some changes

MNarath1 committed:
make it an assoc list

this is basically a typecacheof list not the biggest out there but still unsure if i should keep it that way.

MNarath1 committed:
oops

MNarath1 commented:
i could technically do this with a big istype too but that one would really be slow

MNarath1 committed:
comment

MNarath1 committed:
another batch

MNarath1 committed:
assoc list isn't worth it

MNarath1 commented:
now using only istype means the blocker could be missed but in the end it should still be fine since this is just a fallback option and the assoc list was appearantly more expensive in memory and also kinda processing so i decided to go back to the istype solution

MNarath1 committed:
changing check order

MNarath1 committed:
almost forgot the tables

MNarath1 committed:
changes

MNarath1 committed:
small additions for the possibilty of getting interrupted etc

MNarath1 committed:
don't need 2 identical Comments

MNarath1 committed:
ah whatever

MNarath1 committed:
this is fine

MNarath1 committed:
i am not going to overwrite a whole proc for a single argument!!!

MNarath1 committed:
better way to check for blockers and moving target handling

MNarath1 commented:
Now there are some uhhh concerns with the speed of the spores honestly someone else needs to take a look at this same for how far away you should be able to get before the spore gets its pathfinding canceled so it stops hunting you momentarly.

MNarath1 committed:
comment change
